stdlib/asm/crypto/stark/utils.masm

```diff
@@ -48,3 +48,118 @@ export.validate_inputs
+
+    # z and z^N
+    push.0.0 exec.constants::z_ptr mem_loadw
+    # => [(z_1, z_0)^n, z_1, z_0, alpha1, alpha0, ...]
+
+    exec.constants::get_trace_length_log
+    exec.constants::get_max_cycle_length_log
+    sub
+    # => [log(min_num_cycles), (z_1, z_0)^n, z_1, z_0, alpha1, alpha0, ...] where min_num_cycles = trace_len / max_cycle_len
+
+    dup.4 dup.4
+    # => [z_1, z_0, k, (z_1, z_0)^n, z_1, z_0, alpha1, alpha0, ...] where k = min_num_cycles
+
+    push.1
+    while.true
+        dup.1 dup.1
+        # => [z_1, z_0, z_1, z_0, k, (z_1, z_0)^n, z_1, z_0, alpha1, alpha0, ...] where k = min_num_cycles
+        ext2mul
+        # => [z_exp_1, z_exp_0, k, (z_1, z_0)^n, z_1, z_0, alpha1, alpha0, ...] where k = min_num_cycles
+        dup.2 sub.1 swap.3
+        push.1 neq
+    end
+
+    movup.2 drop
+    # => [z_k_1, z_k_0, (z_1, z_0)^n, z_1, z_0, alpha1, alpha0, ...] where k = min_num_cycles
+
+    exec.constants::get_trace_domain_generator
+    inv
+    # => [g^-1, z_k_1, z_k_0, (z_1, z_0)^n, z_1, z_0, alpha1, alpha0, ...] where k = min_num_cycles
+    movdn.2
+    push.0
+    movdn.2
+    # => [z_k_1, z_k_0, 0, g^-1, (z_1, z_0)^n, z_1, z_0, alpha1, alpha0, ...] where k = min_num_cycles
+
+    exec.constants::get_trace_domain_generator
+    inv dup mul
+    push.0
+    # => [0, g^-2, z_k_1, z_k_0, 0, g^-1, (z_1, z_0)^n, z_1, z_0, alpha1, alpha0, ...] where k = min_num_cycles
+
+    # Save [z_k_0, z_k_1, g^-2, 0]
+    exec.constants::auxiliary_ace_inputs_ptr add.8
+    mem_storew
+    dropw
+    # => [0, g^-1, (z_1, z_0)^n, z_1, z_0, alpha1, alpha0, ...]
+
+    # Save [(z_0, z_1)^n, g^-1, 0]
+    exec.constants::auxiliary_ace_inputs_ptr add.4
+    mem_storew
+    dropw
+
+    # Save [alpha0, alpha1, z_0, z_1]
+    exec.constants::auxiliary_ace_inputs_ptr
+    mem_storew
+    dropw
+    # => [...]
+end
+
+#! Executes the constraints evaluation logic based on arithmetic circuit evaluation.
+#!
+#! Input: [...]
+#! Output: [...]
+export.constraint_evaluation_check
+    # 1) Compute and store at the appropriate memory location the auxiliary inputs needed by
+    #    the arithmetic circuit.
+    exec.set_up_auxiliary_inputs_ace
+    # => [...]
+
+    # 2) Load the circuit description from the advice tape and check that it matches the hardcoded digest.
+    exec.constants::get_arithmetic_circuit_eval_digest
+    adv.push_mapval
+    exec.constants::get_arithmetic_circuit_ptr
+    padw padw padw
+    repeat.12
+        adv_pipe
+        hperm
+    end
+    dropw swapw dropw
+    movup.4 drop
+    assert_eqw
+    # => [...]
+
+    # 3) Set up the inputs to the "arithmetic_circuit_eval" op. Namely:
+    #    1. a pointer to the inputs of the circuit in memory,
+    #    2. the number of inputs to the circuit,
+    #    3. the number of evaluation gates in the circuit.
+    exec.constants::get_arithmetic_circuit_eval_number_eval_gates
+    exec.constants::get_arithmetic_circuit_eval_number_inputs
```

Comment:
Since these are constant, they will not need to be included in the circuit hash right?

Reply:
Yeah, these have the same statues as the digest of the ACE circuit, i.e., there are a constant as the digest of the ACE circuit is a constant and they are hardcoded in the recursive verifier masm file.